cli/src/commands/squash.rs

```diff
@@ -155,21 +222,106 @@ pub(crate) fn cmd_squash(
             ));
         }
         sources = vec![source];
-        destination = parents.pop().unwrap();
-    }
+        pre_existing_destination = Some(parents.pop().unwrap());
+    };
+
+    workspace_command.check_rewritable(sources.iter().chain(&pre_existing_destination).ids())?;
+
+    // prepare the tx description before possibly rebasing the source commits
+    let source_ids: Vec<_> = sources.iter().ids().collect();
+    let tx_description = if let Some(destination) = &pre_existing_destination {
+        format!("squash commits into {}", destination.id().hex())
+    } else {
+        match &source_ids[..] {
+            [] => format!("squash {} commits", source_ids.len()),
+            [id] => format!("squash commit {}", id.hex()),
+            [first, others @ ..] => {
+                format!("squash commit {} and {} more", first.hex(), others.len())
+            }
+        }
+    };
+
+    let mut tx = workspace_command.start_transaction();
+    let destination = if let Some(commit) = pre_existing_destination {
+        commit
+    } else {
+        // create the new destination commit
+        let (parent_ids, child_ids) = compute_commit_location(
+            ui,
+            tx.base_workspace_helper(),
+            args.destination.as_deref(),
+            args.insert_after.as_deref(),
+            args.insert_before.as_deref(),
+            "squash commit",
+        )?;
+        tx.base_workspace_helper()
+            .check_rewritable(sources.iter().ids().chain(child_ids.iter()))?;
+        let parent_commits: Vec<_> = parent_ids
+            .iter()
+            .map(|commit_id| {
+                tx.base_workspace_helper()
+                    .repo()
+                    .store()
+                    .get_commit(commit_id)
+            })
+            .try_collect()?;
+        let merged_tree = merge_commit_trees(tx.repo(), &parent_commits).block_on()?;
+        let mut commit_builder = tx
+            .repo_mut()
+            .new_commit(parent_ids.clone(), merged_tree.id())
+            .detach();
+        let mut description = join_message_paragraphs(&args.message_paragraphs);
+        if !description.is_empty() {
+            commit_builder.set_description(description);
+            description = add_trailers(ui, &tx, &commit_builder)?;
+        }
+        commit_builder.set_description(&description);
+        let commit = commit_builder.write(tx.repo_mut())?;
+        for child_id in &child_ids {
+            let child_commit = tx.repo().store().get_commit(child_id)?;
+            let new_parent_ids = child_commit
+                .parent_ids()
+                .iter()
+                .filter(|id| !&parent_ids.contains(id))
+                .cloned()
+                .chain(std::iter::once(commit.id().clone()))
+                .collect_vec();
+            let new_child_commit =
+                rebase_commit(tx.repo_mut(), child_commit, new_parent_ids).block_on()?;
```

Comment:
Maybe we should do rebase in `.transform_descendats()`.

I think we can use `duplicate_commits(.., &[root], ..)` to insert an empty placeholder commit, but there might be a better function.

Reply:
We still need to update the sources with the new commits if some of the sources have been rewritten.
I've made the code more concise.